README.md
# Hibernate optional query testing

The purpose of this project is to test if Hibernate Reactive can handle optional queries. This idea came by reading https://www.baeldung.com/spring-data-jpa-null-parameters and https://stackoverflow.com/questions/2488930/passing-empty-list-as-parameter-to-jpa-query-throws-error.

**NOTE:** All Hibernate Reactive test are done with Mutiny.

## Formatting

**Prerequisite:** Npm

Formatting is done by prettier and its plugins:

```bash script
npm ci
npx prettier --write .
```

## Running the application

**Prerequisite:** Java, maven and docker (testcontainers)

### Tests

Test are ran either from each subproject or form the parent project with the command below

```bash
mvn clean test --fail-at-end
```

Each test launches a postgres in a testcontainer on a random port and each of them loads the `import.sql` script with the help of Hibernate.


CHANGELOG.md
# Changelog

## Commit 168246d44dce12fc3c401027df79fb15d238df23

Took this refactoring a bit further and also updated all the test with a working Hibernate ORM 6 query that works across all implementations. I needed to define a new parameter for single check `:parameterPresent = FALSE` and list check queries making the solution from https://www.baeldung.com/spring-data-jpa-null-parameters not applicable in Hibernate ORM 6 and the same for https://stackoverflow.com/a/66956214 answer which was the test for for all previous commits so i went with https://stackoverflow.com/a/67179440 suggestion instead, but the https://stackoverflow.com/a/2489053 answer with ceritera queries is probably a better approach for this kind of solution (the current working one feels like a workaround for this).

## Commit ee6d3c3737b126b3464ae0913d9e564215133981

So yeah... the single check queries from baeldung do not work at all here with UUID being null, so i think the only help would be to use the same approach as with `IN :parameter` queries, adding another value to the query ☹. See ee6d3c3737b126b3464ae0913d9e564215133981-failures.md

## Commit 8350d6c3a16b78cd480445cc4dfdcd1011faa5fa

Rename done to new group id and fixed formatting in file from d303fc144a1c10325364f05e11a9661880a94bc5-failures.md

## Commit d303fc144a1c10325364f05e11a9661880a94bc5

Hibernate 5 doing everything awesome again and this is where I tried to just pass in an empty list or null list into a standard JPQL/HQL query and noticed that the `COALESCE(:list) IS NOT NULL` is not needed for optional queries with list, this also caused a rename of the whole project to `<groupId>se.agreedskiing.hibernate.optional</groupId>` in the commits coming after this one. Still Hibernate Reactive or in this case Mutiny can't map UUID and ID to a list query in case of the query `SELECT t FROM Test t WHERE (COALESCE(:list) IS NOT NULL OR t.uu IN (:list))` 🤯, and Hibernate ORM 6 still has problems with an empty list in the optional query.

Next I will try to do optional queries with single item like link from baeldung and start combiding list queries with `AND`.

For the failures check [the commit markdown file](test-failures/d303fc144a1c10325364f05e11a9661880a94bc5-failures.md)

## Commit cc63cdeb5038a41f2e1b673cc22834cefe7d6bc8

This commit works great with Hibernate 5 and replaced immutable list and empty collections.empty with arraylists.

But Hibernate ORM 6 still produces som errors but less.

For the failures check [the commit markdown file](test-failures/cc63cdeb5038a41f2e1b673cc22834cefe7d6bc8-failures.md).

## Commit 00e1cc72480f6de27ccccf62417108c36cd17772

This commit works great with Hibernate 5 but some errors are produced in Hibernate Reactive and Hibernate ORM 6 when doing trying to do the `SELECT t FROM Test t WHERE (COALESCE(:list) IS NOT NULL OR t.id IN (:list))`, `t.uu` or `t.field` where id is a Long, uu is an UUID and field is a String.

So many errors from Hibernate 6 ORM, but wait it does not seem to like Inmutable lists and Collections.empty 😮.

For the failures check [the commit markdown file](test-failures/00e1cc72480f6de27ccccf62417108c36cd17772-failures.md).


hibernate-orm-six/src/test/java/se/agreedskiing/hibernate/optional/hibernate/six/Test.java
package se.agreedskiing.hibernate.optional.hibernate.six;

import jakarta.persistence.*;
import java.util.UUID;

@Entity
@Table(name = "test")
public class Test {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  public Long id;

  public String field; //NOSONAR

  public UUID uu; //NOSONAR
}


hibernate-orm-five/src/test/java/se/agreedskiing/hibernate/optional/hibernate/five/Test.java
package se.agreedskiing.hibernate.optional.hibernate.five;

import java.util.UUID;
import javax.persistence.*;

@Entity
@Table(name = "test")
public class Test {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  public Long id;

  public String field; //NOSONAR

  public UUID uu; //NOSONAR
}


hibernate-orm-six/src/test/java/se/agreedskiing/hibernate/list/hibernate/six/ErrorTexts.java
package se.agreedskiing.hibernate.list.hibernate.six;

public enum ErrorTexts {
  QUERY_FAILED("Query failed with an exception instead of working");

  public final String explenation;

  private ErrorTexts(final String exeplnation) {
    this.explenation = exeplnation;
  }
}


hibernate-reactive/src/main/java/se/agreedskiing/hibernate/list/hibernate/reactive/MutinyApplication.java
package se.agreedskiing.hibernate.list.hibernate.reactive;

import io.smallrye.mutiny.Uni;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import javax.persistence.Persistence;
import org.hibernate.reactive.mutiny.Mutiny;
import org.hibernate.reactive.mutiny.Mutiny.SessionFactory;
import se.agreedskiing.hibernate.list.hibernate.reactive.entities.Test;

public class MutinyApplication {

  public static void main(final String[] args) {
    //NO use currently
  }

  public Uni<List<Test>> listContaingLongs(
    final String persistenceUnit,
    final int port
  ) {
    return queryLongs(persistenceUnit, port, List.of(1L, 2L, 3L));
  }

  public Uni<List<Test>> emptyListOfLongs(
    final String persistenceUnit,
    final int port
  ) {
    return queryLongs(persistenceUnit, port, Collections.emptyList());
  }

  public Uni<List<Test>> nullListOfLongs(
    final String persistenceUnit,
    final int port
  ) {
    return queryLongs(persistenceUnit, port, null);
  }

  public Uni<List<Test>> listContaingUUIDs(
    final String persistenceUnit,
    final int port
  ) {
    return queryUUIDs(
      persistenceUnit,
      port,
      List.of(
        UUID.fromString("40e6215d-b5c6-4896-987c-f30f3678f608"),
        UUID.fromString("6ecd8c99-4036-403d-bf84-cf8400f67836"),
        UUID.fromString("6366d53a-c35c-41b2-90ce-6c43b73490d9"),
        UUID.fromString("37a55e88-5794-48d2-9805-0a44e6edb0b3")
      )
    );
  }

  public Uni<List<Test>> emptyListOfUUIDs(
    final String persistenceUnit,
    final int port
  ) {
    return queryUUIDs(persistenceUnit, port, Collections.emptyList());
  }

  public Uni<List<Test>> nullListOfUUIDs(
    final String persistenceUnit,
    final int port
  ) {
    return queryUUIDs(persistenceUnit, port, null);
  }

  public Uni<List<Test>> listContaingStrings(
    final String persistenceUnit,
    final int port
  ) {
    return queryStrings(
      persistenceUnit,
      port,
      List.of("test1", "test2", "test3", "test4")
    );
  }

  public Uni<List<Test>> emptyListOfStrings(
    final String persistenceUnit,
    final int port
  ) {
    return queryStrings(persistenceUnit, port, Collections.emptyList());
  }

  public Uni<List<Test>> nullListOfStrings(
    final String persistenceUnit,
    final int port
  ) {
    return queryStrings(persistenceUnit, port, null);
  }

  private Uni<List<Test>> queryLongs(
    final String persistenceUnit,
    final int port,
    final List<Long> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery(
            "SELECT t FROM Test t WHERE (COALESCE(:list) IS NOT NULL OR t.id IN (:list))",
            Test.class
          )
          .setParameter("list", values)
          .getResultList()
      );
  }

  private Uni<List<Test>> queryUUIDs(
    final String persistenceUnit,
    final int port,
    final List<UUID> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery(
            "SELECT t FROM Test t WHERE (COALESCE(:list) IS NOT NULL OR t.uu IN (:list))",
            Test.class
          )
          .setParameter("list", values)
          .getResultList()
      );
  }

  private Uni<List<Test>> queryStrings(
    final String persistenceUnit,
    final int port,
    final List<String> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery(
            "SELECT t FROM Test t WHERE (COALESCE(:list) IS NOT NULL OR t.field IN (:list))",
            Test.class
          )
          .setParameter("list", values)
          .getResultList()
      );
  }

  private SessionFactory createSessionFactory(
    final String persistenceUnit,
    final int port
  ) {
    return Persistence
      .createEntityManagerFactory(
        persistenceUnit,
        Map.of(
          "javax.persistence.jdbc.url",
          "jdbc:postgresql://localhost:" + port + "/hibernate_reactive"
        )
      )
      .unwrap(Mutiny.SessionFactory.class);
  }
}


hibernate-reactive/src/main/java/se/agreedskiing/hibernate/optional/hibernate/reactive/Application.java
package se.agreedskiing.hibernate.optional.hibernate.reactive;

import io.smallrye.mutiny.Uni;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import javax.persistence.Persistence;
import org.hibernate.reactive.mutiny.Mutiny;
import org.hibernate.reactive.mutiny.Mutiny.SessionFactory;
import se.agreedskiing.hibernate.optional.hibernate.reactive.entities.Test;

public class Application {

  public static void main(final String[] args) {
    //NO use currently
  }

  public Uni<List<Test>> listContaingLongs(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    final var list = new ArrayList<Long>();
    list.add(1L);
    list.add(2L);
    list.add(3L);
    list.add(4L);
    if (testOptionalQuery) {
      return queryOptionalLongs(persistenceUnit, port, list);
    } else {
      return queryLongs(persistenceUnit, port, list);
    }
  }

  public Uni<List<Test>> emptyListOfLongs(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    final var list = new ArrayList<Long>();
    if (testOptionalQuery) {
      return queryOptionalLongs(persistenceUnit, port, list);
    } else {
      return queryLongs(persistenceUnit, port, list);
    }
  }

  public Uni<List<Test>> nullListOfLongs(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    if (testOptionalQuery) {
      return queryOptionalLongs(persistenceUnit, port, null);
    } else {
      return queryLongs(persistenceUnit, port, null);
    }
  }

  public Uni<List<Test>> listContaingUUIDs(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    final var list = new ArrayList<UUID>();
    list.add(UUID.fromString("40e6215d-b5c6-4896-987c-f30f3678f608"));
    list.add(UUID.fromString("6ecd8c99-4036-403d-bf84-cf8400f67836"));
    list.add(UUID.fromString("6366d53a-c35c-41b2-90ce-6c43b73490d9"));
    list.add(UUID.fromString("37a55e88-5794-48d2-9805-0a44e6edb0b3"));
    if (testOptionalQuery) {
      return queryOptionalUUIDs(persistenceUnit, port, list);
    } else {
      return queryUUIDs(persistenceUnit, port, list);
    }
  }

  public Uni<List<Test>> emptyListOfUUIDs(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    final var list = new ArrayList<UUID>();
    if (testOptionalQuery) {
      return queryOptionalUUIDs(persistenceUnit, port, list);
    } else {
      return queryUUIDs(persistenceUnit, port, list);
    }
  }

  public Uni<List<Test>> nullListOfUUIDs(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    if (testOptionalQuery) {
      return queryOptionalUUIDs(persistenceUnit, port, null);
    } else {
      return queryUUIDs(persistenceUnit, port, null);
    }
  }

  public Uni<List<Test>> listContaingStrings(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    final var list = new ArrayList<String>();
    list.add("test1");
    list.add("test2");
    list.add("test3");
    list.add("test4");
    if (testOptionalQuery) {
      return queryOptionalStrings(persistenceUnit, port, list);
    } else {
      return queryStrings(persistenceUnit, port, list);
    }
  }

  public Uni<List<Test>> emptyListOfStrings(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    final var list = new ArrayList<String>();
    if (testOptionalQuery) {
      return queryOptionalStrings(persistenceUnit, port, list);
    } else {
      return queryStrings(persistenceUnit, port, list);
    }
  }

  public Uni<List<Test>> nullListOfStrings(
    final String persistenceUnit,
    final int port,
    final boolean testOptionalQuery
  ) {
    if (testOptionalQuery) {
      return queryOptionalStrings(persistenceUnit, port, null);
    } else {
      return queryStrings(persistenceUnit, port, null);
    }
  }

  private Uni<List<Test>> queryOptionalLongs(
    final String persistenceUnit,
    final int port,
    final List<Long> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery(
            "SELECT t FROM Test t WHERE (COALESCE(:list) IS NOT NULL OR t.id IN (:list))",
            Test.class
          )
          .setParameter("list", values)
          .getResultList()
      );
  }

  private Uni<List<Test>> queryLongs(
    final String persistenceUnit,
    final int port,
    final List<Long> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery("SELECT t FROM Test t WHERE t.id IN (:list)", Test.class)
          .setParameter("list", values)
          .getResultList()
      );
  }

  private Uni<List<Test>> queryOptionalUUIDs(
    final String persistenceUnit,
    final int port,
    final List<UUID> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery(
            "SELECT t FROM Test t WHERE (COALESCE(:list) IS NOT NULL OR t.uu IN (:list))",
            Test.class
          )
          .setParameter("list", values)
          .getResultList()
      );
  }

  private Uni<List<Test>> queryUUIDs(
    final String persistenceUnit,
    final int port,
    final List<UUID> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery("SELECT t FROM Test t WHERE t.uu IN :list", Test.class)
          .setParameter("list", values)
          .getResultList()
      );
  }

  private Uni<List<Test>> queryOptionalStrings(
    final String persistenceUnit,
    final int port,
    final List<String> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery(
            "SELECT t FROM Test t WHERE (COALESCE(:list) IS NOT NULL OR t.field IN (:list))",
            Test.class
          )
          .setParameter("list", values)
          .getResultList()
      );
  }

  private Uni<List<Test>> queryStrings(
    final String persistenceUnit,
    final int port,
    final List<String> values
  ) {
    return createSessionFactory(persistenceUnit, port)
      .withSession(s ->
        s
          .createQuery(
            "SELECT t FROM Test t WHERE t.field IN (:list)",
            Test.class
          )
          .setParameter("list", values)
          .getResultList()
      );
  }

  private SessionFactory createSessionFactory(
    final String persistenceUnit,
    final int port
  ) {
    return Persistence
      .createEntityManagerFactory(
        persistenceUnit,
        Map.of(
          "javax.persistence.jdbc.url",
          "jdbc:postgresql://localhost:" + port + "/hibernate_reactive"
        )
      )
      .unwrap(Mutiny.SessionFactory.class);
  }
}
